src/ai/ai.cpp

```diff
@@ -104,17 +104,32 @@ void update_ai_general_status(sys::state& state) {
 	}
 }
 
+static void get_alliance_targets_by_adjacency(sys::state& state, dcon::nation_id n, dcon::nation_id adj, std::vector<dcon::nation_id>& alliance_targets) {
+	for(auto nb : state.world.nation_get_nation_adjacency(adj)) {
+		auto other = nb.get_connected_nations(0) != adj ? nb.get_connected_nations(0) : nb.get_connected_nations(1);
+		if(other != n && other.get_is_player_controlled() == false && !(other.get_overlord_as_subject().get_ruler()) && !nations::are_allied(state, n, other) && !military::are_at_war(state, other, n) && ai_will_accept_alliance(state, other, n))
+			alliance_targets.push_back(other.id);
+	}
+}
+
 void form_alliances(sys::state& state) {
 	static std::vector<dcon::nation_id> alliance_targets;
 
 	for(auto n : state.world.in_nation) {
 		if(!n.get_is_player_controlled() && n.get_ai_is_threatened() && !(n.get_overlord_as_subject().get_ruler())) {
 			alliance_targets.clear();
 
-			for(auto nb : n.get_nation_adjacency()) {
-				auto other = nb.get_connected_nations(0) != n ? nb.get_connected_nations(0) : nb.get_connected_nations(1);
-				if(other.get_is_player_controlled() == false && !(other.get_overlord_as_subject().get_ruler())  && !nations::are_allied(state, n, other) && !military::are_at_war(state, other, n) && ai_will_accept_alliance(state, other, n))
-					alliance_targets.push_back(other.id);
+			// Adjacency with us
+			get_alliance_targets_by_adjacency(state, n.id, n.id, alliance_targets);
+			// Adjacency with rival (useful for e.x, Chile allying Paraguay to fight bolivia)
+			get_alliance_targets_by_adjacency(state, n.id, n.get_ai_rival(), alliance_targets);
```

Comment:
Running this twice could end up with duplicate nations in the list

Reply:
handled below with std::unique

Reply:
Also, the rival is not always set. You need to check whether the rival exists in the first place

Reply:
you are still letting duplicates be added